database/test_database.go

```diff
@@ -868,3 +881,157 @@ func TestCompactNoPanic(t *testing.T, db Database) {
+func benchmarkSetup(b *testing.B, count int, size int) ([][]byte, [][]byte) {
+	keys := make([][]byte, count)
+	values := make([][]byte, count)
+	for i := 0; i < b.N; i++ {
+		keyBytes := make([]byte, b.N)
+		valueBytes := make([]byte, size)
+		rand.Read(keyBytes)
+		rand.Read(valueBytes)
+		keys[i], values[i] = keyBytes, valueBytes
+	}
+	return keys, values
+}
+
+// BenchmarkGet measures the time it takes to get an operation from a database.
+func BenchmarkGet(b *testing.B, db Database, name string, count int, size int) {
+	keys, values := benchmarkSetup(b, count, size)
+
+	b.Run(fmt.Sprintf("%s_%d_%dbytes_get", name, count, size), func(b *testing.B) {
+		// Writes random values of size _size_ to the database
+		for i := 0; i < b.N; i++ {
+			if err := db.Put(keys[i%count], values[i%count]); err != nil {
+				b.Fatalf("Unexpected error in Put %s", err)
+			}
+		}
+		b.ResetTimer()
+
+		// Reads b.N values from the db
+		for i := 0; i < b.N; i++ {
+			if _, err := db.Get(keys[i%count]); err != nil {
+				b.Fatalf("Unexpected error in Get %s", err)
+			}
+		}
+	})
+}
+
+// BenchmarkPut measures the time it takes to write an operation to a database.
+func BenchmarkPut(b *testing.B, db Database, name string, count int, size int) {
+	keys, values := benchmarkSetup(b, count, size)
+
+	b.Run(fmt.Sprintf("%s_%d_%dbytes_put", name, count, size), func(b *testing.B) {
+		// Writes b.N values to the db
+		for i := 0; i < b.N; i++ {
+			if err := db.Put(keys[i%count], values[i%count]); err != nil {
+				b.Fatalf("Unexpected error in Put %s", err)
+			}
+		}
+	})
+}
+
+// BenchmarkDelete measures the time it takes to delete a (k, v) from a database.
+func BenchmarkDelete(b *testing.B, db Database, name string, count int, size int) {
+	keys, values := benchmarkSetup(b, count, size)
+
+	b.Run(fmt.Sprintf("%s_%d_%dbytes_delete", name, count, size), func(b *testing.B) {
+		// Writes random values of size _size_ to the database
+		for i := 0; i < b.N; i++ {
+			if err := db.Put(keys[i%count], values[i%count]); err != nil {
+				b.Fatalf("Unexpected error in Put %s", err)
+			}
+		}
+		b.ResetTimer()
+
+		// Deletes b.N values from the db
+		for i := 0; i < b.N; i++ {
+			if err := db.Delete(keys[i%count]); err != nil {
+				b.Fatalf("Unexpected error in Delete %s", err)
+			}
+		}
+	})
+}
+
+// BenchmarkBatchPut measures the time it takes to batch write.
+func BenchmarkBatchPut(b *testing.B, db Database, name string, count int, size int) {
+	keys, values := benchmarkSetup(b, count, size)
+
+	b.Run(fmt.Sprintf("%s_%d_%dbytes_write_batch", name, count, size), func(b *testing.B) {
+		batch := db.NewBatch()
+		if batch == nil {
+			b.Fatalf("db.NewBatch returned nil")
+		}
+		for i := 0; i < b.N; i++ {
+			if err := batch.Put(keys[i%count], values[i%count]); err != nil {
+				b.Fatalf("Unexpected error in db.Put: %s", err)
+			}
+			if err := batch.Write(); err != nil {
+				b.Fatalf("Unexpected error in batch.Write: %s", err)
+			}
```

Comment:
Was it intended for the batch size to grow here with each loop iteration?

Reply:
The rationale here was that this might show how cheap / expensive it is to use `batch.Write`. However this is clearly not the right way to benchmark batched writes. I thought perhaps having a separate `BenchmarkInterface` for batching would be better than hardcoding a fixed set of iterations within an inner loop.

Reply:
I was mainly commenting that the batch is going to grow with each `Put` because the batch is never `Reset`. So the first `Write` is going to write a single kv pair. The next write will write two kv pairs, ect.

Reply:
Ah right, in that case I added `batch.Reset` to the batch benchmark after `Write`. The intent was not for the batch to grow.